### PR TITLE
[kinetic] improve cross-compile behaviour

### PR DESCRIPTION
**Pull request by dominiquehunziker**
Fix from openembedded's [meta-ros](https://github.com/bmwcarit/meta-ros/pull/386) layer to generate pkgConfig.cmake files which can be used also on the target platform to compile packages.

When cross-compiling the pkgConfig.cmake contains absolute paths from the host platform which makes it impossible to use the pkgConfig.cmake files to compile packages in the target platform. This PR addresses this by stripping the `CMAKE_FIND_ROOT_PATH` prefixes from the relevant variables before they are used to generate the pkgConfig.cmake file. In a second step CMake's `find_*` functions are used in the pkgConfig.cmake file to handle `CMAKE_FIND_ROOT_PATH` correctly when locating the previously stripped variables.

### CODE REVIEW

File: cmake/catkin_package.cmake
Comment (dirk-thomas) [inline cmake/catkin_package.cmake]: This looks wrong to me. E.g. if the root path contains `/some/path` this would break paths which start with `/some/pathplusmore`.

Comment (dominiquehunziker) [inline cmake/catkin_package.cmake]: If I understand this correctly, this should be fixed by doing something like

``` cmake
string(REGEX REPLACE "^${_path}/" "/" _ele ${_ele})
```

(just off the top of my head, without testing)

Comment (dirk-thomas) [inline cmake/catkin_package.cmake]: No, that would make the result an absolute path. Only the regex needs the trailing slash.

**Comment by dirk-thomas** [comment]
While this might be helpful for cross compilation I am worried about the potential side effects for the common use case. Maybe it would be better to only conditionally enable this behavior?

File: cmake/catkin_package.cmake
Comment (dominiquehunziker) [inline cmake/catkin_package.cmake]: This is the intended behavior; the result should be again an absolute path, but without the root path. Otherwise the generated pkgConfig.cmake file would see a relative path and would therefore be treated differently in the include logic. However, we want this path to be treated as an absolute path in the target filesystem.

**Comment by dominiquehunziker** [comment]
Sure that can be done; I didn't make it explicit, because AFAIK the CMAKE_FIND_ROOT_PATH would only be set in a toolchain file for cross-compiling. I haven't encountered a usecase of CMAKE_FIND_ROOT_PATH where it is not used in the context of cross compilation.

So, would you suggest to guard this code with CMAKE_CROSSCOMPILING or should some other option be introduced somewhere?

**Comment by dirk-thomas** [comment]
My worry is that not all the changes are depending on `CMAKE_FIND_ROOT_PATH` to be set. I will add some more comments inline.

File: cmake/catkin_package.cmake
Comment (dirk-thomas) [inline cmake/catkin_package.cmake]: OK, sounds good.

File: cmake/catkin_package.cmake
Comment (dirk-thomas) [inline cmake/catkin_package.cmake]: This should be a function instead of a macro. The only variable being changed in the parent scope should be the passed in list. That can easily be achieved with `set(... PARENT_SCOPE)`.

Why use two different terms? The function could be named `catkin_strip_root_path` to match what it does.

The function should also be moved to its own file and contain a docblock.

File: cmake/catkin_package.cmake
Comment (dirk-thomas) [inline cmake/catkin_package.cmake]: Instead of a `LIST` argument the function should have one positional argument like `function(catkin_strip_root_path var)`. Also `var` is the common name for this use. Instead of `_fn` the prefix should be named `ARG` (for consistency with other code).

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in]: Please avoid changing the semantic of a variable in the middle. It makes it more difficult to read and understand the logic.

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in]: You are using `find_path` here to check if the absolute path is available under any of the root paths, right?

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in]: In case my previous assumption about `find_path` is right this error message seems weird. We had an absolute path before, why would the message only mention the folder name? Also even if the name is not found in any root path it should always be available under the original absolute path. So why should this happen anyway?

If I understand it correctly you are using `find_path` here to find the same absolute path nested under a potential root path, right?

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in]: This looks like an unrelated bug fix.

**Comment by dirk-thomas** [comment]
As long as with an empty `CMAKE_FIND_ROOT_PATH` the logic is pretty much identical I don't think we need an option. It might be good to get feedback from other people doing cross compilation on this, too.

File: cmake/catkin_strip_root_path.cmake
Comment (dirk-thomas) [inline cmake/catkin_strip_root_path.cmake]: The parameter is not the actual list but the variable name containing the list:

```
:param var: the input / output variable name
:type var: string
```

File: cmake/catkin_strip_root_path.cmake
Comment (dirk-thomas) [inline cmake/catkin_strip_root_path.cmake]: Should this `break()` after the first match?

File: cmake/catkin_strip_root_path.cmake
Comment (dirk-thomas) [inline cmake/catkin_strip_root_path.cmake]: Since the function has its own local scope please avoid prefixing variables with an underscore.

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in]: The name `FIND_XXX_LOCATION_OPTION` makes it look like an "official" variable of CMake. Since it is only a custom local variable please use `_find_root_path_both` instead.

File: cmake/catkin_strip_root_path.cmake
Comment (dominiquehunziker) [reply cmake/catkin_strip_root_path.cmake]: Yes. Multiple replacements don't make sense; this would mean there is a root filesystem nested in another root filesystem or there is a false positive match. I guess this requires a string comparison of new and old value or is there a better way for this?

File: cmake/templates/pkgConfig.cmake.in
Comment (dominiquehunziker) [reply cmake/templates/pkgConfig.cmake.in]: Yes, this is unrelated. However, according to the doc the important part is the `-NOTFOUND` suffix and this is just a cosmetic fix.

File: cmake/templates/pkgConfig.cmake.in
Comment (dominiquehunziker) [reply cmake/templates/pkgConfig.cmake.in]: Ok. I was wondering though whether it is a good idea to allow CMake to search for header files & libraries in the host filesystem. As I mentioned as a comment it is necessary to find catkin files, but if this rule is applied to standard includes like `/usr/include` this might have undesired effects. I thought about explicitly searching the catkin workspace and not the host filesystem, but in case workspaces are chained this will get complicated fast.

File: cmake/catkin_strip_root_path.cmake
Comment (dirk-thomas) [reply cmake/catkin_strip_root_path.cmake]: Just calling `break()` after this line should be sufficient imo. I don't think we need to check for nesting.

Comment (dominiquehunziker) [reply cmake/catkin_strip_root_path.cmake]: If this would be the case then there would be no need for the foreach loop, no? This would always just try to strip the first element in CMAKE_FIND_ROOT_PATH if there is an unconditional break.

Comment (dirk-thomas) [reply cmake/catkin_strip_root_path.cmake]: You are right. The break should only happen after a successful replace. And yes, that will require a comparison of the old and new value.

Maybe the regex should go into a variable and the check with `if(... MATCHES ...)` and inside the block perform the substitution and break?

File: cmake/catkin_strip_root_path.cmake
Comment (dirk-thomas) [inline cmake/catkin_strip_root_path.cmake:13]: Please also check that no other unexpected arguments are being passed (like here: https://github.com/ros/catkin/blob/063f057fa0f83bbde23da322c3d37f6e54bbbb27/cmake/catkin_package.cmake#L107-L109).

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in:116]: Please don't refer to the project maintainer directly but use the existing logic which first tries to reference the web urls of the package: https://github.com/ros/catkin/blob/063f057fa0f83bbde23da322c3d37f6e54bbbb27/cmake/templates/pkgConfig.cmake.in#L97-L103

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in:86]: Should this be set to an empty value in all other cases? Since the variable is being used later unconditionally.

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [inline cmake/templates/pkgConfig.cmake.in:162]: This will change the current behavior (when not using `CMAKE_CROSSCOMPILING`). Before `NO_CMAKE_FIND_ROOT_PATH` was always passed. With this patch it is not passed anymore. Can you please explain why that should be the case.

**Review by dirk-thomas — COMMENTED** [review]
I am just mentioning a few users here which have contributed in the past to cross compile related functionality with the hope that they can provide feedback for this patch: @stonier @herbrechtsmeier @mightyCelu @mikepurvis @housel

File: cmake/templates/pkgConfig.cmake.in
Comment (mxsrc) [reply cmake/templates/pkgConfig.cmake.in:162]: Do you remember what the reason was to use `NO_CMAKE_FIND_ROOT_PATH` here? Because it should only be set in crosscompilation builds anyway and in such a build using it is pretty much a necessity.

**Comment by mxsrc** [comment]
Without having tested the changes it looks like it should work.

I wonder though why the lookup for files is done both in the usual root as well as in the `CMAKE_FIND_ROOT_PATH` since this can lead to issues when e.g. libraries are present on both systems.

EDIT: After reading the changes this makes sense to me now as I explained in the answer above.

File: cmake/templates/pkgConfig.cmake.in
Comment (dirk-thomas) [reply cmake/templates/pkgConfig.cmake.in:162]: I would assume `NO_CMAKE_FIND_ROOT_PATH` is being used (since ed2ac4ffe16b469a89bc700a3149cbc6ac33aae2) to prevent that the library is being found anywhere else than in the specific path `${path}`.

Comment (mxsrc) [reply cmake/templates/pkgConfig.cmake.in:162]: I see. Removing it should not break anything since it is only relevant when cross-compiling, in which case we actually want to include the `CMAKE_FIND_ROOT_PATH`. Also, the current change-set would break when building for install-space if the parameter is kept.

Comment (dirk-thomas) [reply cmake/templates/pkgConfig.cmake.in:162]: Changing the argument in this PR does change the logic where the library might be found. Right? As far as I understand the code the argument is being passed so that the library is only found in the locations specified by `PKG_CONFIG_LIB_PATHS`. So removing the argument would change that in non cross-compile scenarios. Why should that be done? What is the intention of changing this line? Can you describe a case where the current behavior is wrong?

Comment (mxsrc) [reply cmake/templates/pkgConfig.cmake.in:162]: Yes, the change of the argument changes the lookup location but only if `CMAKE_FIND_ROOT_PATH` is set, which is only the case in cross-compiled builds (afaik, a quick survey in the #cmake IRC confirmed this). In a regular build there will be no change.

The current behavior is only wrong when it comes to cross-compiled builds. In such builds, you usualy have two filesystems available, the host-machine's and the target's (under a prefix in the host filesystem). The person who builds can specify whether the libraries/files required for the build should be looked up in the host- or target-filesystem), or both. Usually you'd want to look up libraries in the rootfs since they need to have the correct architecture.

The current behavior overrides this and forces the lookup to be only on the host-filesystem. This will break for every build since the library will not be found (if the host does not have the library installed to the same location) or fail when compiling (if installed to the same location) since the architectures won't match.

This is even unrelated from the introduced flag, which is required when not installing but rather building to devel-space where we want to find a target-architecture-library with a host-path.

Comment (dirk-thomas) [reply cmake/templates/pkgConfig.cmake.in:162]: Thank you for the detailed explanation. I think I got it now why the change makes sense.

**Comment by mateus-amarante** [comment]
I'm  trying to cross compile a ROS package, but I'm having a similar problem to the one related in this topic: 

https://answers.ros.org/question/207608/error-while-cross-compiling-rospkg-for-arm/#

Might this error be related to this pull request?

**Comment by mxsrc** [comment]
I don't think the error is related to this PR. The issue you mention is caused by librt.so being missing or in a different location than expected.

File: cmake/templates/pkgConfig.cmake.in
Comment (mxsrc) [reply cmake/templates/pkgConfig.cmake.in:86]: It looks to me like the empty value will be correct for all usages. In the uncoditional case, the value will either be empty in a regular build or, for a crosscompiled build, the one being set by the developer will be used.

**Review by mxsrc — APPROVED** [review]

**Comment by dirk-thomas** [comment]
@dominiquehunziker I am still waiting for my pending comments to be addressed before re-reviewing this and considering to merge it.

**Comment by dirk-thomas** [comment]
Closing due to inactivity.

**Pull request closed without merging**
